File: solutions/0940-fruit-into-baskets/solution.js
/**
 * @param {number[]} fruits
 * @return {number}
 */
var totalFruit = function(fruits) {
    let currentFruits = [];
    const lastSeen = {};
    let max = 0

    let l = 0;
    let r = 0;

    while(r < fruits.length){
        let current = fruits[r];
        lastSeen[current] = r;


        if(currentFruits.length < 2 && !currentFruits.includes(current)){
            currentFruits.push(current);
        }

        if(!currentFruits.includes(current)){
            currentFruits = currentFruits.sort((a,b) => lastSeen[b] - lastSeen[a]);

            let oldFruit = currentFruits.pop();
            l = lastSeen[oldFruit] + 1;
            currentFruits.push(current);
        }

        max = Math.max(max, r - l + 1);
        r++;
    }


    return max
};

File: solutions/0940-fruit-into-baskets/solution.py
class Solution:
    def totalFruit(self, fruits: List[int]) -> int:
        types = {}
        left = 0
        total = 0
        for i in range(len(fruits)):
            next_tree = fruits[i]

            types[next_tree] = i
            if len(types) > 2:
                for t in types:
                    if t != next_tree and t != fruits[i - 1]:
                        left = types[t] + 1
                        del types[t]
                        break

            total = max(total, i - left + 1)

        return total 
